### Task 8.1
* Name:LI WAN
* Student Number:223718804
* E-mail:s223718804@deakin.edu.au
* Course:SIT720
---

In [6]:
# Q1 Download Electrical Grid Stability Simulated Datadatasets. Classify "Electrical Grid Stability Simulated Data" classes using KNN. Use the same data splitting and performance metrics that you have used in previous week (week 7, Q-2). Report your findings including comparison of results with week 7 
import pandas as pd
local_path = r'E:\2-学习\1-Deakin\24-T1\SIT720\Task\Data_for_UCI_named.csv'
df=pd.read_csv(local_path, comment='#')
df

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
...        ...       ...       ...       ...       ...       ...       ...   
9995  2.930406  9.487627  2.376523  6.187797  3.343416 -0.658054 -1.449106   
9996  3.392299  1.274827  2.954947  6.894759  4.349512 -1.663661 -0.952437   
9997  2.364034  2.842030  8.776391  1.008906  4.299976 -1.380719 -0.943884   
9998  9.631511  3.994398  2.757071  7.821347  2.514755 -0.966330 -0.649915   
9999  6.530527  6.781790  4.349695  8.673138  3.492807 -1.390285 -1.532193   

            p4        g1        g2        g3        g4      stab     stabf  
0    -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1    -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2    -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3    -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4    -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable  
...        ...       ...       ...       ...       ...       ...       ...  
9995 -1.236256  0.601709  0.779642  0.813512  0.608385  0.023892  unstable  
9996 -1.733414  0.502079  0.567242  0.285880  0.366120 -0.025803    stable  
9997 -1.975373  0.487838  0.986505  0.149286  0.145984 -0.031810    stable  
9998 -0.898510  0.365246  0.587558  0.889118  0.818391  0.037789  unstable  
9999 -0.570329  0.073056  0.505441  0.378761  0.942631  0.045263  unstable  

[10000 rows x 14 columns]

In [10]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder

# Split the data into features (X) and target variable (y)
X = df.drop(columns=['stabf'])
y = df['stabf']

# Split the data into training and testing sets (80% train)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Standardize the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Initialize the KNN classifier
knn = KNeighborsClassifier()# default as 'Euclidean' distance

# Train the KNN model on the training data
knn.fit(X_train_scaled, y_train)
#knn.fit(X_train,  y_train)

# Make predictions on the test data
y_pred = knn.predict(X_test_scaled)
#y_pred = knn.predict(X_test)

# Calculate performance metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro')
recall = recall_score(y_test, y_pred, average='macro')
f1 = f1_score(y_test, y_pred, average='macro')
conf_matrix = confusion_matrix(y_test, y_pred)

# Perform cross-validation
cv_scores = cross_val_score(knn, X, y, cv=5)  # Perform 5-fold cross-validation

# Cross-validation
mean_cv_score = cv_scores.mean()

# Print performance metrics
print("Performance Metrics:")
print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print(f"F1-score: {f1:.2f}")
print("Confusion Matrix:")
print(conf_matrix)
print("Cross-Validation Scores:", cv_scores)
print("Mean Cross-Validation Score:", mean_cv_score)

Performance Metrics:
Accuracy: 0.94
Precision: 0.94
Recall: 0.93
F1-score: 0.93
Confusion Matrix:
[[ 636   88]
 [  36 1240]]
Cross-Validation Scores: [0.7765 0.787  0.7895 0.7855 0.778 ]
Mean Cross-Validation Score: 0.7832999999999999


Based on the provided results, we can compare the performance of the KNN (K-Nearest Neighbors) and SVM (Support Vector Machine) with a linear kernel for the task of detecting "Electrical Grid Stability Simulated Data".

KNN:
- Accuracy: 0.94
- Precision: 0.94
- Recall: 0.93
- F1-score: 0.93
- Mean Cross-Validation Score: 0.7833

SVM with linear kernel:
- Accuracy: 0.92
- Precision: 0.92
- Recall: 0.91
- F1-score: 0.91
- Mean Cross-Validation Score: 0.92

From these results, we can observe the following:

1. **Accuracy**: KNN has a slightly higher accuracy of 0.94 compared to 0.92 for SVM with a linear kernel.

2. **Precision**: KNN and SVM have the same precision of 0.94 and 0.92, respectively.

3. **Recall**: KNN has a slightly higher recall of 0.93 compared to 0.91 for SVM.

4. **F1-score**: KNN has a slightly higher F1-score of 0.93 compared to 0.91 for SVM.

5. **Confusion Matrix**: Both models have a similar confusion matrix, with KNN having slightly fewer false positives (88 vs. 86) and fewer false negatives (36 vs. 72) compared to SVM.

6. **Cross-Validation Score**: SVM with a linear kernel has a higher mean cross-validation score of 0.92 compared to 0.7833 for KNN. This suggests that SVM with a linear kernel might generalize better to unseen data.

Overall, both KNN and SVM with a linear kernel performed well on this task, with KNN having slightly better overall performance metrics (accuracy, precision, recall, and F1-score) on the test set. However, SVM with a linear kernel has a higher mean cross-validation score, indicating better generalization to unseen data.

It's important to note that the choice between KNN and SVM (or any other algorithm) depends on the specific problem, the characteristics of the data, and the trade-offs between different performance metrics that are most important for the task at hand. Additionally, hyperparameter tuning and feature engineering can significantly impact the performance of both algorithms.

In [9]:
# Q2  Load Electrical Grid Stability Simulated Data and create classification model using DT algorithm using 50-50% and 80-20% data splitting methods. Compare performances of these two models and explain impact of difference in data splitting on the performances of the model.
from sklearn.tree import DecisionTreeClassifier
# Separate features and target variable
X = df.drop('stabf', axis=1)
y = df['stabf']

# 50-50% data splitting
X_train_50, X_test_50, y_train_50, y_test_50 = train_test_split(X, y, test_size=0.5, random_state=42)

# 80-20% data splitting
X_train_80, X_test_80, y_train_80, y_test_80 = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize Decision Tree Classifier with optimized hyperparameters
dt_classifier = DecisionTreeClassifier(max_depth=8, min_samples_split=20, min_samples_leaf=10, max_features='sqrt', random_state=42)

# Train the model using 50-50% data splitting method
dt_classifier.fit(X_train_50, y_train_50)

# Make predictions on the test set using 50-50% data splitting method
y_pred_50 = dt_classifier.predict(X_test_50)

# Evaluate the model using different metrics for 50-50% data splitting method
accuracy_50 = accuracy_score(y_test_50, y_pred_50)
precision_50 = precision_score(y_test_50, y_pred_50, average='macro')
recall_50 = recall_score(y_test_50, y_pred_50, average='macro')
f1_50 = f1_score(y_test_50, y_pred_50, average='macro')
conf_matrix_50 = confusion_matrix(y_test_50, y_pred_50)

# Train the model using 80-20% data splitting method
dt_classifier.fit(X_train_80, y_train_80)

# Make predictions on the test set using 80-20% data splitting method
y_pred_80 = dt_classifier.predict(X_test_80)

# Evaluate the model using different metrics for 80-20% data splitting method
accuracy_80 = accuracy_score(y_test_80, y_pred_80)
precision_80 = precision_score(y_test_80, y_pred_80, average='macro')
recall_80 = recall_score(y_test_80, y_pred_80, average='macro')
f1_80 = f1_score(y_test_80, y_pred_80, average='macro')
conf_matrix_80 = confusion_matrix(y_test_80, y_pred_80)

# Print the performance metrics for both data splitting methods
print("Performance Metrics for 50-50% Data Splitting Method:")
print(f"Accuracy: {accuracy_50:.2f}")
print(f"Precision: {precision_50:.2f}")
print(f"Recall: {recall_50:.2f}")
print(f"F1-score: {f1_50:.2f}")
print("Confusion Matrix:")
print(conf_matrix_50)

print("\nPerformance Metrics for 80-20% Data Splitting Method:")
print(f"Accuracy: {accuracy_80:.2f}")
print(f"Precision: {precision_80:.2f}")
print(f"Recall: {recall_80:.2f}")
print(f"F1-score: {f1_80:.2f}")
print("Confusion Matrix:")
print(conf_matrix_80)

Performance Metrics for 50-50% Data Splitting Method:
Accuracy: 0.97
Precision: 0.97
Recall: 0.97
F1-score: 0.97
Confusion Matrix:
[[1733   62]
 [  90 3115]]

Performance Metrics for 80-20% Data Splitting Method:
Accuracy: 0.92
Precision: 0.91
Recall: 0.92
F1-score: 0.91
Confusion Matrix:
[[ 639   54]
 [ 109 1198]]


Based on the provided performance metrics, we can compare the two models and analyze the impact of the difference in data splitting on their performances:

1. **Accuracy**:
   - 50-50% Data Splitting: Accuracy = 0.97
   - 80-20% Data Splitting: Accuracy = 0.92

The model trained with 50-50% data splitting has a higher accuracy of 0.97 compared to 0.92 for the model trained with 80-20% data splitting.

2. **Precision**:
   - 50-50% Data Splitting: Precision = 0.97
   - 80-20% Data Splitting: Precision = 0.91

The model trained with 50-50% data splitting has a higher precision of 0.97, meaning it has fewer false positives compared to the model trained with 80-20% data splitting, which has a precision of 0.91.

3. **Recall**:
   - 50-50% Data Splitting: Recall = 0.97
   - 80-20% Data Splitting: Recall = 0.92

The model trained with 50-50% data splitting has a higher recall of 0.97, indicating that it has fewer false negatives compared to the model trained with 80-20% data splitting, which has a recall of 0.92.

4. **F1-score**:
   - 50-50% Data Splitting: F1-score = 0.97
   - 80-20% Data Splitting: F1-score = 0.91

The F1-score, which is the harmonic mean of precision and recall, is higher for the model trained with 50-50% data splitting (0.97) compared to the model trained with 80-20% data splitting (0.91).

5. **Confusion Matrix**:
   - 50-50% Data Splitting: The confusion matrix shows fewer false positives (62) and fewer false negatives (90) compared to the 80-20% data splitting model.
   - 80-20% Data Splitting: The confusion matrix shows more false positives (54) and more false negatives (109) compared to the 50-50% data splitting model.

The difference in performance between the two models can be attributed to the difference in data splitting. The 50-50% data splitting method provides a more balanced training and testing set, which can lead to better generalization and higher performance metrics.

On the other hand, the 80-20% data splitting method allocates a smaller portion of the data for testing (20%), which can result in a less representative test set and potentially lower performance metrics. Additionally, with a smaller test set, the performance metrics may be more sensitive to outliers or imbalances in the data.

It's important to note that the choice of data splitting method depends on the specific problem, the size of the dataset, and the trade-off between model complexity and generalization performance. In general, a larger test set (e.g., 80-20% or even 70-30% splitting) is preferred for a more reliable estimate of the model's performance on unseen data. However, if the dataset is small, a more balanced split like 50-50% or cross-validation techniques may be necessary to make the most of the available data.

In [44]:
# Q3 Create two more KNN-based classification models using the dataset used in Q1 by varying distance metrics such as using  cityblock and manhattan. Report the performances of the developed models including Q1 and explain the similarity or differences if any.
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Standardize the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Initialize and train KNN models with different distance metrics
knn_cityblock = KNeighborsClassifier(n_neighbors=5, metric='cityblock')
knn_manhattan = KNeighborsClassifier(n_neighbors=5, metric='manhattan')

knn_cityblock.fit(X_train_scaled, y_train)
knn_manhattan.fit(X_train_scaled, y_train)

# Make predictions
y_pred_cityblock = knn_cityblock.predict(X_test_scaled)
y_pred_manhattan = knn_manhattan.predict(X_test_scaled)

# Evaluate performance
accuracy_cityblock = accuracy_score(y_test, y_pred_cityblock)
accuracy_manhattan = accuracy_score(y_test, y_pred_manhattan)

precision_cityblock = precision_score(y_test, y_pred_cityblock, average='macro')
precision_manhattan = precision_score(y_test, y_pred_manhattan, average='macro')

recall_cityblock = recall_score(y_test, y_pred_cityblock, average='macro')
recall_manhattan = recall_score(y_test, y_pred_manhattan, average='macro')

f1_cityblock = f1_score(y_test, y_pred_cityblock, average='macro')
f1_manhattan = f1_score(y_test, y_pred_manhattan, average='macro')

conf_matrix_cityblock = confusion_matrix(y_test, y_pred_cityblock)
conf_matrix_manhattan = confusion_matrix(y_test, y_pred_manhattan)

print("KNN with Cityblock Distance Metric:")
print(f"Accuracy: {accuracy_cityblock:.2f}")
print(f"Precision: {precision_cityblock:.2f}")
print(f"Recall: {recall_cityblock:.2f}")
print(f"F1-score: {f1_cityblock:.2f}")
print("Confusion Matrix:")
print(conf_matrix_cityblock)

print("\nKNN with Manhattan Distance Metric:")
print(f"Accuracy: {accuracy_manhattan:.2f}")
print(f"Precision: {precision_manhattan:.2f}")
print(f"Recall: {recall_manhattan:.2f}")
print(f"F1-score: {f1_manhattan:.2f}")
print("Confusion Matrix:")
print(conf_matrix_manhattan)

KNN with Cityblock Distance Metric:
Accuracy: 0.94
Precision: 0.94
Recall: 0.93
F1-score: 0.93
Confusion Matrix:
[[ 633   60]
 [  58 1249]]

KNN with Manhattan Distance Metric:
Accuracy: 0.94
Precision: 0.94
Recall: 0.93
F1-score: 0.93
Confusion Matrix:
[[ 633   60]
 [  58 1249]]


Based on the provided performance metrics, we can analyze the similarities and differences between the models:

KNN with Euclidean Distance Metrics:

Accuracy: 0.94
Precision: 0.94
Recall: 0.93
F1-score: 0.93


KNN with Cityblock Distance Metric:

Accuracy: 0.94 (same as Euclidean)
Precision: 0.94 (same as Euclidean)
Recall: 0.93 (same as Euclidean)
F1-score: 0.93 (same as Euclidean)
Confusion Matrix: Similar to Euclidean, but with slightly fewer false positives (60 vs. 88) and more false negatives (58 vs. 36)


KNN with Manhattan Distance Metric:

Accuracy: 0.94 (same as Euclidean and Cityblock)
Precision: 0.94 (same as Euclidean and Cityblock)
Recall: 0.93 (same as Euclidean and Cityblock)
F1-score: 0.93 (same as Euclidean and Cityblock)
Confusion Matrix: Identical to the Cityblock distance metric



Similarities:

All three models (Euclidean, Cityblock, and Manhattan) have the same overall accuracy, precision, recall, and F1-score of 0.94, 0.94, 0.93, and 0.93, respectively.
The confusion matrices are similar, with a slight difference in the number of false positives and false negatives between Q1 and the Cityblock/Manhattan distance metrics.

Differences:

The only notable difference is in the confusion matrices, where the Cityblock and Manhattan distance metrics have slightly fewer false positives (60 vs. 88 in Q1) and more false negatives (58 vs. 36 in Euclidean) compared to the baseline Euclidean model.

Based on these results, it seems that changing the distance metric from the default Euclidean to Cityblock or Manhattan did not significantly impact the overall performance of the KNN models. All three models performed similarly well, with only minor differences in the distribution of false positives and false negatives.
It's worth noting that the choice of distance metric in KNN can impact the model's performance, especially when dealing with different types of data distributions or feature scales. However, in this particular case, the differences are relatively minor, and all three models achieved similar and reasonably good performance.